clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/loadbalancer/cache/AzureLoadBalancerCachingAgent.groovy

```diff
@@ -75,30 +91,233 @@ class AzureLoadBalancerCachingAgent implements CachingAgent, AccountAware {
+            } else {
+              // Found a load balancer that has been deleted since last time was read from Azure cloud
+              loadBalancer = null
+            }
+          }
+        }
+
+        if (loadBalancer) {
+          data.add(buildCacheData(loadBalancer))
+        }
+      }
+
+      log.info("Caching ${data.size()} items in ${agentType}")
+
+      return new DefaultCacheResult(
+        [(Keys.Namespace.AZURE_LOAD_BALANCERS.ns): data],
+        [(Keys.Namespace.AZURE_ON_DEMAND.ns): evictions])
+
+    } else {
+      if (updatedLoadBalancer) {
+        // This is an OnDemand update/edit request for a given load balancer resource
+        // Attempt to add entry into the OnDemand respective cache
+        if (updateCache(providerCache, updatedLoadBalancer, "OnDemandUpdated")) {
+          CacheData data = buildCacheData(updatedLoadBalancer)
+
+          log.info("Caching 1 OnDemand updated item in ${agentType}")
+          return new DefaultCacheResult([(Keys.Namespace.AZURE_LOAD_BALANCERS.ns): [data]])
+        } else {
+          return null
+        }
+      }
 
-    def loadBalancers = creds.networkClient.getLoadBalancersAll(creds, region)
-
-    List<CacheData> data = loadBalancers.collect() { AzureLoadBalancerDescription loadBalancer ->
-      Map<String, Object> attributes = [loadbalancer: loadBalancer]
-      new DefaultCacheData(Keys.getLoadBalancerKey(azureCloudProvider,
-                                                   loadBalancer.loadBalancerName,
-                                                   loadBalancer.loadBalancerName,
-                                                   loadBalancer.appName?:"none",
-                                                   loadBalancer.cluster?:"none",
-                                                   region,
-                                                   accountName),
-        attributes,
-        [:])
+      if (evictedLoadBalancer) {
+        // This is an OnDemand delete request for a given load balancer resource
+        // Attempt to add entry into the OnDemand respective cache
+        if (updateCache(providerCache, evictedLoadBalancer, "OnDemandEvicted")) {
+          log.info("Caching 1 OnDemand evicted item in ${agentType}")
+          return new DefaultCacheResult([(Keys.Namespace.AZURE_LOAD_BALANCERS.ns): []])
+        } else {
+          return null
+        }
+      }
     }
 
-    log.info("Caching ${data.size()} items in ${agentType}")
-    new DefaultCacheResult([(Keys.Namespace.AZURE_LOAD_BALANCERS.ns): data])
+    return new DefaultCacheResult([(Keys.Namespace.AZURE_LOAD_BALANCERS.ns): []])
+  }
+
+  // Update current cache only if the current entry is a new OnDemand request or if the current entry is more recent
+  //  than the cache entry
+  private Boolean updateCache(ProviderCache providerCache, AzureLoadBalancerDescription loadBalancer, String onDemandCacheType) {
+    Boolean foundUpdatedOnDemandLB = false
```

Comment:
We will get back to it once the application/cluster/serve group caching agents are implemented because there might give us more opportunities to consolidate some of the code shared currently by load balancer and security groups.